runtime/android/core/src/org/xwalk/core/XWalkActivity.java

```diff
@@ -5,16 +5,560 @@
+                    }
+
+                    int statusIdx = cursor.getColumnIndex(DownloadManager.COLUMN_STATUS);
+                    int status = cursor.getInt(statusIdx);
+                    if (status == DownloadManager.STATUS_FAILED ||
+                            status == DownloadManager.STATUS_SUCCESSFUL) {
+                        return status;
+                    } else if (status == DownloadManager.STATUS_PAUSED) {
+                        if (++pausedCount == MAX_PAUSED_COUNT) return status;
+                    } else if (status == DownloadManager.STATUS_RUNNING) {
+                        if (++runningCount == MAX_RUNNING_COUNT) return status;
+                    }
+                }
+            }
+
+            return 0;
+        }
+
+        @Override
+        protected void onProgressUpdate(Integer... progress) {
+            Log.d("XWalkActivity", "Download progress: " + progress[0] + "/" + progress[1]);
+            mProgressDialog.setProgress(progress[0]);
+            mProgressDialog.setMax(progress[1]);
+            mProgressDialog.setIndeterminate(false);
+        }
+
+        @Override
+        protected void onCancelled(Integer result) {
+            Log.d("XWalkActivity", "Download cancelled");
+            mProgressDialog.dismiss();
+            mXWalkActivity.onXWalkLibraryCancelled();
+        }
+
+        @Override
+        protected void onPostExecute(Integer result) {
+            mProgressDialog.dismiss();
+
+            if (result == DownloadManager.STATUS_SUCCESSFUL) {
+                Log.d("XWalkActivity", "Download finished");
+                Uri uri = mDownloadManager.getUriForDownloadedFile(mDownloadId);
+                mXWalkActivity.installXWalkLibrary(uri);
+            } else {
+                String errMsg = mXWalkActivity.getString(R.string.download_failed_message);
+                if (result == DownloadManager.STATUS_FAILED) {
+                    Query query = new Query().setFilterById(mDownloadId);
+                    Cursor cursor = mDownloadManager.query(query);
+                    if (cursor != null && cursor.moveToFirst()) {
+                        int reasonIdx = cursor.getColumnIndex(DownloadManager.COLUMN_REASON);
+                        int reason = cursor.getInt(reasonIdx);
+                        if (reason == DownloadManager.ERROR_DEVICE_NOT_FOUND) {
+                            errMsg = mXWalkActivity.getString(
+                                    R.string.download_failed_device_not_found) ;
+                        } else if (reason == DownloadManager.ERROR_INSUFFICIENT_SPACE) {
+                            errMsg = mXWalkActivity.getString(
+                                    R.string.download_failed_insufficient_space);
+                        }
+                    }
+                } else if (result == DownloadManager.STATUS_PAUSED) {
+                } else if (result == DownloadManager.STATUS_RUNNING) {
```

Comment:
May specify special message for STATUS_PAUSED later. Currently, just use default message.